Security Reporting guidelines (#277)
nodejs/package-maintenance: docs/drafts/security-guidelines.md

## What

Guidelines for responsible reporting and disclosure of security vulnerabilities.

## Why

Defining a process of reporting and disclosing security vulnerabilities helps maintainers protect projects and their users.

## Security policy guidelines

### Publishing the policy

Each project should document its security policy in an easily accessible place.

Here are several common ways of publishing security policies:

1. Security section on project website, e.g. `https://PROJECTSITE/security`.
1. [security.txt](https://securitytxt.org/) file on project website.

> 1. [Security policy in a GitHub repository](https://help.github.com/en/github/managing-security-vulnerabilities/adding-a-security-policy-to-your-repository).

Reviewer: i think this should be the first listed.

Reviewer 2: Agreed!  Many projects do not have websites but all the ones we are concerned with have repos (mostly on githbu).  And right below this should be the `SECURITY.md` file, as that is the second most common in my expirence.

1. `SECURITY.md` or `SECURITY` file in the root folder of project repository (if hosted outside of GitHub).
1. Security section of the project `README` file.

### Reporting a security issue

> Project security policy should describe a process for reporting vulnerabilities. It is encouraged that the reporting process is confidential to allow maintainers to triage and resolve the vulnerability in private before disclosing it to the public.

Reviewer 2: ```suggestion
Project security policy should describe a process for reporting vulnerabilities. It is strongly encouraged that the reporting process is confidential to allow maintainers to triage and resolve the vulnerability in private before disclosing it to the public.
```

> Here are several common ways of reporting vulnerabilities:

Reviewer 3: I think the first bullet should be `Follow the instruction in the security policy of the project`

Author: The goal of this guideline document (in my mind) is to help project maintainers establish security vulnerability reporting policies for their projects.

With this in mind, I assumed the reader of this document to be looking to create a policy from scratch or adopt a policy from another project or the Ecosystem Security WG.

Is this what you were after?

Reviewer 3: Yes, I misunderstood the subject
Thank for the clarification

Author: No worries, this is useful feedback. I will make the audience clearer in the intro section.

> 1. Designated email address, e.g. `security@PROJECTSITE`.

Reviewer 2: This reads weird to me, maybe just use `@example.com`?

> 1. [Node.js Ecosystem Security WG](https://github.com/nodejs/security-wg) responsible disclosure program on [HackerOne](https://hackerone.com/nodejs-ecosystem).

Reviewer 3: We could say also: `if a module doesn't have a secury policy, you can always fallow the [third party reporting](https://github.com/nodejs/security-wg/blob/master/processes/third_party_vuln_process.md)`

Author: So again, what you are saying is that this document should be more prescriptive and should present a way for maintainers to pick their way of handling vulnerability reports?

Reviewer 3: Sorry again, in my mind the subject was a user that wanted to report a security bug.

The policy should clearly specify when confidential reporting is not supported and generally available bug tracking system should be used.

### Responsible disclosure

> Public disclosure should follow release of a patch that addresses a vulnerability.

Reviewer 2: Could we add something about working with the maintainer to do the public disclosure?  I have seen many times that the report to these tools is not quite right in it's version ranges, or that the description is not quite right, and this causes confusion and work for maintainers.

It would be much better if the maintainers were involved so that they are aware and can collaborate on what their users will see.

Reviewer 4: The main issue that occurs with version ranges is instead of asking the author they sometimes just assume it affected all previous versions and/or there is confusion if there were multiple release lines patched

Author: Wouldn't this be something to put into the security policy itself, rather than this guidelines document?

My reasoning here is that this guidelines document will mostly be read and used by package maintainers, not the security research community.

Reviewer: If package maintainers are aware that this is something they need to discuss explicitly with the security researcher, however, that might help things.

Reviewer 2: Agreed, I think the goal of putting it here would be to make it more expected and known for all parties involved.

Author: @dougwilson @wesleytodd does the new paragraph about collaboration address this comment as well?

Reviewer 2: It does as far as I am concerned :)

Here are the recommended ways to disclose a vulnerability:

1. [Create a maintainer advisory on GitHub](https://help.github.com/en/github/managing-security-vulnerabilities/creating-a-maintainer-security-advisory).
1. [Report a vulnerability to npm](https://www.npmjs.com/advisories/report).
1. Submit a pull request to [vulnerability database](https://github.com/nodejs/security-wg/blob/master/processes/vuln_db.md) maintained by [Node.js Ecosystem Security WG](https://github.com/nodejs/security-wg).

Note that vulnerabilities reported to the responsible disclosure program on [HackerOne](https://hackerone.com/nodejs-ecosystem) will be automatically disclosed by HackerOne staff and the Ecosystem Security WG triage team members.

## Examples

> Here are several examples of short and useful security policies:

Reviewer 3: I would improve this section because the use cases are different and we could help to choose the user:

- Node.js Ecosystem Security WG Template -> the template valid for every nodejs package
- Express Security Policies and Procedures -> support custom (email) and the common Node.js Ecosystem Security WG Template
- Node-RED Security Policy -> prefer a custom process via email

[In Fastify we aligned to the common Node.js Ecosystem Security WG Template](https://github.com/fastify/fastify/blob/master/SECURITY.md)

So we have all the case:
- if a module doesn't have a security policy (wins the Node.js Ecosystem Security WG Template)
- if a module supports a custom process + the common one
- if a module prefer a custom process
- if a module state the common one

Author: We definitely could change this section from `Examples` to `How to craft your security policy` depending on how prescriptive we want this guidance to be.

@ljharb @mhdawson @wesleytodd @dougwilson @Eomm given that most maintainers are probably not security experts, do you feel we should make this more prescriptive rather than just suggesting options?

Reviewer 5: I think the discussion you have with the options, followed by providing our "recommendation" along with a template that could be used would be good. It would provide some background info and then provide our recommended approach that can be easily followed unless the maintainer has specific requirements/reasons to do otherwise.

Reviewer 2: I would guess that most maintainers have a process they would like to follow despite not always being "security experts" to minimize the impact of supposed "security experts" reporting publicly.  So in that light, I agree that options with a recommendation is good.

Also, if the maintainer does not expect to support their package with prompt security updates that is fine, but they should state that in their policy.  And we should add a section about that.

Author: This section aims to cover this:

https://github.com/MarcinHoppe/package-maintenance/blob/vuln-reporting-guidelines/docs/drafts/security-guidelines.md#releasing-security-patches

@wesleytodd Do you this it should be expanded or rewritten? Is there a particular concern that is not addressed?

Reviewer 2: Expanded I think.  The goal is to get the reporter to work hand in hand with the maintainer.  So they should send a full report of the disclosure plan (text and versions included) to the maintainer before doing anything else.  This is specifically the thing which has caused issued in the past with Express, reporters opening a disclosure with bad information.

Author: I understand the concern now. I will expand this section to guide towards a more collaborative disclosure.

Author: @wesleytodd does https://github.com/MarcinHoppe/package-maintenance/blob/vuln-reporting-guidelines/docs/drafts/security-guidelines.md#releasing-security-patches help address your concern?

Reviewer 2: Great addition!  Thanks!

1. [Node.js Ecosystem Security WG](https://github.com/nodejs/security-wg) responsible disclosure program on [HackerOne](https://hackerone.com/nodejs-ecosystem).

The policy should clearly specify when confidential reporting is not supported and generally available bug tracking system should be used.

### Responsible disclosure

Public disclosure should follow release of a patch that addresses a vulnerability.

Here are the recommended ways to disclose a vulnerability:

1. [Create a maintainer advisory on GitHub](https://help.github.com/en/github/managing-security-vulnerabilities/creating-a-maintainer-security-advisory).
1. [Report a vulnerability to npm](https://www.npmjs.com/advisories/report).
1. Submit a pull request to [vulnerability database](https://github.com/nodejs/security-wg/blob/master/processes/vuln_db.md) maintained by [Node.js Ecosystem Security WG](https://github.com/nodejs/security-wg).

Note that vulnerabilities reported to the responsible disclosure program on [HackerOne](https://hackerone.com/nodejs-ecosystem) will be automatically disclosed by HackerOne staff and the Ecosystem Security WG triage team members.

## Examples

> Here are several examples of short and useful security policies:

Reviewer 3: ```suggestion
Here are several examples of short and useful security policies that fit different maintainers' needs:
```

> - [Node.js Ecosystem Security WG Template](https://github.com/nodejs/security-wg/blob/master/processes/responsible_disclosure_template.md)

Reviewer 3: ```suggestion
- [Node.js Ecosystem Security WG Template](https://github.com/nodejs/security-wg/blob/master/processes/responsible_disclosure_template.md): the common security policy that all packages may follow
```

> - [Express Security Policies and Procedures](https://github.com/expressjs/express/security/policy)

Reviewer 3: ```suggestion
- [Express Security Policies and Procedures](https://github.com/expressjs/express/security/policy): support a customized workflow and the Node.js Ecosystem Security process
```

> - [Node-RED Security Policy](https://github.com/node-red/node-red/security/policy)

Reviewer 3: ```suggestion
- [Node-RED Security Policy](https://github.com/node-red/node-red/security/policy): has defined a dedicated security reporting workflow
```

> - [Fastify Security Policy](https://github.com/fastify/fastify/blob/master/SECURITY.md) (refers to the Node.s Ecosystem Security WG process for reporting vulnerabilities)

Reviewer 3: ```suggestion
- [Fastify Security Policy](https://github.com/fastify/fastify/blob/master/SECURITY.md): adopt the Node.s Ecosystem Security WG process for reporting vulnerabilities
```

## What

Guidelines for responsible reporting and disclosure of security vulnerabilities.

## Why

Defining a process of reporting and disclosing security vulnerabilities helps maintainers protect projects and their users.

## Security policy guidelines

### Publishing the policy

Each project should document its security policy in an easily accessible place.

Here are several common ways of publishing security policies:

1. [Security policy in a GitHub repository](https://help.github.com/en/github/managing-security-vulnerabilities/adding-a-security-policy-to-your-repository).
1. `SECURITY.md` or `SECURITY` file in the root folder of project repository (if hosted outside of GitHub).

> 1. Security section on project website, e.g. `https://example.com/security`.

Reviewer 6: ```suggestion
1. Security webpage on project website, e.g. `https://example.com/security`.
```

Do we want to suggest only a section or a an entire page for security? I think page makes sense given the example URL.

> 1. [security.txt](https://securitytxt.org/) file on project website.

Reviewer 6: ```suggestion
1. [security.txt](https://securitytxt.org/) file on project website, e.g. `https://example.com/.well-known/security.txt`
```

Might as well add an example here
